native/protocol/src/x3dh.rs

```diff
@@ -0,0 +1,364 @@
+use crate::aead::{decrypt_data, encrypt_data, AeadError};
+use crate::bundle::*;
+use chacha20poly1305::{
+    aead::{KeyInit, Payload},
+    ChaCha20Poly1305,
+};
+use ed25519_dalek::{Signature, SigningKey, VerifyingKey};
+use hkdf::Hkdf;
+use serde::{Deserialize, Serialize};
+use sha2::Sha256;
+use thiserror::Error;
+use x25519_dalek::{
+    PublicKey as X25519PublicKey, ReusableSecret as X25519ReusableSecret,
+    StaticSecret as X25519StaticSecret,
+};
+
+// See https://signal.org/docs/specifications/x3dh/ for an explanation of the below variable names
+// and functions of the X3DH protocol.
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SignedPreKey {
+    pub pre_key: X25519PublicKey,
+    pub signature: Signature,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SignedPreKeys {
+    pub pre_keys: Vec<X25519PublicKey>,
+    pub signature: Signature,
+}
+
+pub struct X3DHSendKeyAgreement {
+    pub ek: X25519PublicKey,
+    pub sk: [u8; 32],
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Message {
+    pub sender_identity: String,
+    pub sender_ik: VerifyingKey,
+    pub ek: X25519PublicKey,
+    pub opk: Option<X25519PublicKey>,
+    pub ciphertext: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PreKeyBundle {
+    pub ik: VerifyingKey,
+    pub opk: Option<X25519PublicKey>,
+    pub spk: SignedPreKey,
+}
+
+// KDF(KM) represents 32 bytes of output from the HKDF algorithm [3] with inputs:
+//    HKDF input key material = F || KM, where KM is an input byte sequence containing secret key material, and F is a byte sequence containing 32 0xFF bytes if curve is X25519, and 57 0xFF bytes if curve is X448. F is used for cryptographic domain separation with XEdDSA [2].
+//    HKDF salt = A zero-filled byte sequence with length equal to the hash output length.
+//    HKDF info = An ASCII string identifying the application.
+fn kdf(km: &[u8]) -> [u8; 32] {
+    let salt = [0; 32];
+    let f = [0xFF, 32];
+    let ikm = [&f, km].concat();
+    let hk = Hkdf::<Sha256>::new(Some(&salt), &ikm);
+    let mut okm = [0u8; 32];
+    hk.expand(b"Brongnal", &mut okm).unwrap();
+    okm
+}
+
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum X3DHError {
+    #[error("Signature failed to validate.")]
+    SignatureValidation,
+    #[error("Aead routine failed.")]
+    Aead(#[from] AeadError),
+}
+
+// If the bundle does not contain a one-time prekey, she calculates:
+//    DH1 = DH(IKA, SPKB)
```

Comment:
This might be silly, but could you provide a glossary of the acronyms used at the top. I'm having to tab over to the specification page regularly to read this.

Reply:
Great idea!